File: samples/app-complete-sample/nodejs/server/dialogs/teams/UpdateCardMsgDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { CardFactory, ActionTypes, MessageFactory } = require('botbuilder');
const { WaterfallDialog, ComponentDialog } = require('botbuilder-dialogs');
const UPDATECARDMESSAGE = 'UpdateCardMessage';

class UpdateCardMsgDialog extends ComponentDialog {
    constructor(id, conversationDataAccessor) {
        super(id);
        this.conversationDataAccessor = conversationDataAccessor;
        // Define the conversation flow using a waterfall model.
        this.addDialog(new WaterfallDialog(UPDATECARDMESSAGE, [
            this.beginUpdateCardMsgDialog.bind(this),
        ]));
    }

    async beginUpdateCardMsgDialog(stepContext) {
        var currentState = await this.conversationDataAccessor.get(stepContext.context, {});

        if (currentState.lastDialogKey == "UpdateCardMsgSetupDialog" && currentState.activityId != null) 
        {
            const buttons = [
                { type: ActionTypes.MessageBack, title: 'MSDN', value: 'update card message', text: "update card message" },
            ]; 

            const CardImage = CardFactory.images(["https://sec.ch9.ms/ch9/7ff5/e07cfef0-aa3b-40bb-9baa-7c9ef8ff7ff5/buildreactionbotframework_960.jpg"])
            const card = CardFactory.heroCard('Updated card ', CardImage,
                buttons, { subtitle: "This card is updated now" });

            const message = MessageFactory.attachment(card);
            message.id = stepContext.context._activity.replyToId;
            await stepContext.context.updateActivity(message);
            await stepContext.context.sendActivity("Message updated - see above");
            currentState.lastDialogKey = "UpdateCardMsgDialog";
        }
        else {
            currentState.lastDialogKey = "UpdateCardMsgDialog";
            await stepContext.context.sendActivity("No message to update.");
        }

        return await stepContext.endDialog();
    }
}

exports.UpdateCardMsgDialog = UpdateCardMsgDialog;

File: samples/TeamsSDK/Archived/app-complete-sample/csharp/AppCompleteSample/src/dialogs/examples/teams/UpdateCardMsgDialog.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Connector;
using Microsoft.Bot.Schema;
using AppCompleteSample;
using AppCompleteSample.src.dialogs;
using AppCompleteSample.Utility;
using Microsoft.Extensions.Options;
using System.Threading.Tasks;
using System.Threading;
using System;
using System.Collections.Generic;

namespace AppCompleteSample.Dialogs
{
    /// <summary>
    /// This is update card dialog class. Main purpose of this class is to update the card, if user has already setup the card message from below dialog file
    /// microsoft-teams-sample-complete-csharp\template-bot-master-csharp\src\dialogs\examples\teams\updatecardmsgsetupdialog.cs
    /// </summary>
    public class UpdateCardMsgDialog : ComponentDialog
    {
        protected readonly IStatePropertyAccessor<RootDialogState> _conversationState;

        private readonly IOptions<AzureSettings> azureSettings;

        public int updateCounter;
        public UpdateCardMsgDialog(IStatePropertyAccessor<RootDialogState> conversationState, IOptions<AzureSettings> azureSettings) : base(nameof(UpdateCardMsgDialog))
        {
            this._conversationState = conversationState;
            this.azureSettings = azureSettings;
            InitialDialogId = nameof(WaterfallDialog);
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
            {
                BeginUpdateCardMsgDialogAsync,
            }));
        }

        private async Task<DialogTurnResult> BeginUpdateCardMsgDialogAsync(
            WaterfallStepContext stepContext,
            CancellationToken cancellationToken = default(CancellationToken))
        {
            if (stepContext == null)
            {
                throw new ArgumentNullException(nameof(stepContext));
            }
            var existingState = await this._conversationState.GetAsync(stepContext.Context, () => new RootDialogState());
            if (!string.IsNullOrEmpty(stepContext.Context.Activity.ReplyToId))
            {
                IMessageActivity activity = stepContext.Context.Activity;

                updateCounter = TemplateUtility.ParseUpdateCounterJson((Activity)activity);

                var updatedMessage = CreateUpdatedMessage(stepContext);
                ConnectorClient client = new ConnectorClient(new Uri(stepContext.Context.Activity.ServiceUrl), this.azureSettings.Value.MicrosoftAppId, this.azureSettings.Value.MicrosoftAppPassword);

                try
                {
                    ResourceResponse resp = await client.Conversations.UpdateActivityAsync(stepContext.Context.Activity.Conversation.Id, stepContext.Context.Activity.ReplyToId, (Activity)updatedMessage);
                    await stepContext.Context.SendActivityAsync(Strings.UpdateCardMessageConfirmation);
                }
                catch (Exception ex)
                {
                    await stepContext.Context.SendActivityAsync(Strings.ErrorUpdatingCard + ex.Message);
                }
            }
            else
            {
                await stepContext.Context.SendActivityAsync(Strings.NoMsgToUpdate);
            }

            //Set the Last Dialog in Conversation Data
            var currentState = await this._conversationState.GetAsync(stepContext.Context, () => new RootDialogState());
            currentState.LastDialogKey = Strings.LastDialogSetupUpdateCard;
            await this._conversationState.SetAsync(stepContext.Context, currentState);

            return await stepContext.EndDialogAsync(null, cancellationToken);
        }

        #region Create Updated Card Message
        private IMessageActivity CreateUpdatedMessage(WaterfallStepContext context)
        {
            var message = context.Context.Activity;
            if (message.Attachments != null)
            {
                message.Attachments = null;
            }
            var attachment = CreateUpdatedCardAttachment();
            message.Attachments = new List<Attachment>() { attachment };
            return message;
        }
        private Attachment CreateUpdatedCardAttachment()
        {
            return new HeroCard
            {
                Title = Strings.UpdatedCardTitle + " " + updateCounter,
                Subtitle = Strings.UpdatedCardSubTitle,
                Images = new List<CardImage> { new CardImage(this.azureSettings.Value.BaseUri + " /public/assets/computer_person.jpg") },
                Buttons = new List<CardAction>
                {
                   new CardAction(ActionTypes.MessageBack, Strings.UpdateCardButtonCaption, value: "{\"updateKey\": \"" + ++updateCounter + "\"}", text: DialogMatches.UpdateCard)
                }
            }.ToAttachment();
        }
        #endregion
    }
}